docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,405 @@
+multi-signatures and attach these to the block instead of including one
+signature per validator.
+
+#### Block Verification
+
+Currently, we verify each validator signature using the public key associated
+with that validator.  With signature aggregation, verification of blocks would
+not verify many signatures individually, but would instead check the (single)
+multi-signature using the public keys stored by the validator. This would also
+require a mechanism for indicating which validators are included in the
+aggregated signature.
+
+#### IBC Relaying
+
+IBC would no longer need to transmit a large set of signatures when
+updating state. These state updates do not happen for every IBC packet, only
+when changing an IBC light client's view of the counterparty chain's state.
+General [IBC packets][ibc-packet] only contain enough information to correctly
+route the data to the counterparty chain.
+
+IBC does persist commit signatures to the chain in these `MsgUpdateClient`
+message when updating state. This message would no longer need the full set
+of unique signatures and would instead only need one signature for all of the
+data in the header.
+
+Adding BLS signatures would create a new signature type that must be
+understood by the IBC module and by the relayers. For some operations, such
+as state updates, the set of data written into the chain and received by the
+IBC module could be slightly smaller.
+
+## Discussion
+
+### What are the proposed benefits to aggregated signatures?
+
+#### Reduce Block Size
+
+At the moment, a commit contains a 64-byte (512-bit) signature for each validator
+that voted for the block. For the Cosmos Hub, which has 175 validators in the
+active set, this amounts to about 11 KiB per block. That gives an upper bound of
+around 113 GiB over the lifetime of the chain's 10.12M blocks. (Note, the hub has
+increased the number of validators in the active set over time so the total
+signature size over the history of the chain is likely somewhat less than that).
+
+Signature aggregation would only produce two signatures for the entire block.
+One for the yeas and one for the nays. Each BLS aggregated signature is 48
+bytes, per the [IETF standard of BLS signatures][bls-ietf-ecdsa-compare].
+Over the lifetime of the same cosmos hub chain, that would amount to about 1
+GB, a savings of 112 GB. While that is a large factor of reduction it's worth
+bearing in mind that, at [GCP's cost][gcp-storage-pricing] of $.026 USD per GB,
+that is a total savings of around $2.50 per month.
+
+#### Reduce Signature Creation and Verification Time
+
+From the [IETF draft standard on BLS Signatures][bls-ietf], BLS signatures can be
+created in 370 microseconds and verified in 2700 microseconds. Our current
+[Ed25519 implementation][voi-ed25519-perf] lists a 27.5 microsecond signature creation time
+and 5.19 milliseconds to perform batch verification on 128 signatures, which is
```

Comment:
A few notes:

 - The numbers in that document are only really intended to be used for comparison between the primitives I benchmarked due to ` i7-10510U @ 1.80GHz with turbo disabled`, which cripples an already not-amazing mobile part for consistency.

 - Tendermint will use the "Precomputed verification" optimization (or at least, it did when I upstreamed the library switch), because the time-memory tradeoff is worth it in a consensus context (~3.91ms).  (And the blst go benchmark likewise pre-aggregates the public key).

A quick and dirty unscientific test on a Ryzen 5600X (with PBO enabled) looks like this:
```
BenchmarkExpanded/VerifyBatchOnly/128-12          	     662	   2250318 ns/op	  192704 B/op	      17 allocs/op
BenchmarkVerifyFastAggregateUncompressedMinSig/100_G1-12       	    1094	   1166317 ns/op
```

So my nitpicking is somewhat moot, since BLS signature verification is still faster. but the gap isn't quite as big, though the more interesting (to me) parts about using BLS signatures is the size/bandwidth savings.

Reply:
Thank you so much for that notes @Yawning ! It may be worth then re-running these benchmarks on my machine to compare more directly with the numbers here for BLS signatures.

Reply:
Performed this benchmark and noted the findings in this document. I also benchmarked key _aggregation_ times using _blst_ on different numbers of keys and determined that the cost of aggregating a key in a `bls` scheme appears to be expensive.